### DISCUSSION TITLE
Team2 (hostile) NPC AI completely break if the location has a team/faction

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#13818] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods.

### What happened?

If the location doesn't have `outpostteam="None"` and is owned by I assume FriendlyNPC team or a faction, the AI of NPCs in team2 breaks: Enemies flee you without attacking. 

I assume this is to prevent Team2 from shooting at the outpost? But humans don't actually shoot handheld weapons at walls

![image](https://github.com/FakeFishGames/Barotrauma/assets/73229309/5bb7c864-1e1c-484a-a57a-b9be40f66f79)

# IF YOU SPAWN THE ENEMIES WITH FRIENDLY TEAM ID and THEN CHANGE THEIR TEAM, THE AI IS FINE FOR SOME REASON 

```xml
      <Character identifier="soldierseparatists" from="de_separatistscombat" teamid="FriendlyNpc" requirekill="true" />
      
    <ScriptedEvent identifier="takeoverteamtest" >
      <TagAction criteria="humanprefabidentifier:soldierseparatists" tag="defenders" ContinueIfNoTargetsFound="true" />
      <NPCChangeTeamAction npctag="defenders" teamtag="Team2" />
    </ScriptedEvent> 
```

# IF YOU SPAWN THE ENEMIES WITH teamid="Team2" OR LEAVE IT BLANK, THE AI BREAKS
    
This is a problem for any event/mission that uses Team2 to turn NPCs hostile. Otherwise you have to loop a combataction event to circumvent the problem

Example of  modded mission that is basically using the abandonedoutpostassassinate mission, but in a city 

https://github.com/FakeFishGames/Barotrauma/assets/73229309/acb81caf-8ec2-42f1-b0b1-352537ca73c3

```xml
  <OutpostDestroyMission name="Coalition Station-Takeover" identifier="cotakeover_separatistoutpost" tags="outposttakeovermission" type="OutpostDestroy" requireCorrectLocationType="true" locations="city" reward="5500" difficulty="4" commonness="0" allowretry="true" BlockLocationTypeChanges="true">
    <Reputation identifier="separatists" amount="-20" />
    <Reputation identifier="coalition" amount="10" />
    <Icon texture="Content/UI/MissionIcons.png" sourcerect="0,512,256,256" color="214,50,48" />
    <Characters>
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" multiplayeronly="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" multiplayeronly="true" />
      <Character identifier="soldierseparatists" from="de_separatistscombat" requirekill="true" multiplayeronly="true" />
    </Characters>
    <TriggerEvent state="0" delay="1" eventidentifier="cooutposttakeoverevent" campaignonly="true" />
    <LocationType identifier="city" />
    <message header="missionheader0.coalitiontakeover" text="missionmessage0.coalitiontakeover" />
  </OutpostDestroyMission>
```

### Reproduction steps

1. Delete `outpostteam="None"` from abandonedoutpost locationtype.xml
2. Start an abandoned outpost mission with human enemies
3. AI breaks

### Bug prevalence

Happens every time I play

_No response_

### Version

v1.4.4.1 (Blood in the Water Update)

**Comment by ubertpendragon** [comment]
You can see  the jailbreak event devolve into security inspection inception (changes a bot to team 2)

https://github.com/FakeFishGames/Barotrauma/assets/73229309/abb7fb1d-0f35-455f-81d9-58c4bffe0368

**Comment by itchyOwl** [comment]
Is this a new issue in the latest update?

**Reply by ubertpendragon** [reply]
I don't think so, we had this type of mission in the mod for a long time only now it's been reworked and it relied on a combataction loop previously

The security inspection definitely is new

**Comment by Regalis11** [comment]
I'm not entirely sure I understand this correctly.

> If the location doesn't have outpostteam="None" and is owned by I assume FriendlyNPC team or a faction, the AI of NPCs in team2 breaks: Enemies flee you without attacking.

So you mean, if you have an outpost that belongs to "FriendlyNPCs", and you're spawning NPCs belonging to Team2 inside it, they flee instead of actively fighting you? I'm not entirely sure I'd consider this a bug: it's intentional that by default NPCs flee instead of fighting when they encounter hostiles inside a hostile outpost (unless they've been explicitly ordered to fight enemies). I think this could be addressed by changing the location to belong to Team2 instead, which should make the NPCs consider it "their outpost" and defend against intruders.

**Reply by ubertpendragon** [reply]
I don't understand either, but as said in repro steps, the ai only works if:

- location has outpostteam="None", NPCs spawned as Team2
- npcs spawned as "FriendlyNPC", switched to Team2 via event


It is by all means a bug that openly hostile npcs, flee upon seeing you in the video

> changing the location to belong to Team2 instead

The point is to allow Team2 to simply function in any outpost regardless, as they do when you change their team via event. 

The AI is capable of functioning if you change their team via event properly, the bug is that the AI is broken if they are spawned as Team2. I'm not sure why AI would flee in the first place at all

**Comment by UnhappyWithMyUsername** [comment]
> I'm not entirely sure I understand this correctly.
> 
> > If the location doesn't have outpostteam="None" and is owned by I assume FriendlyNPC team or a faction, the AI of NPCs in team2 breaks: Enemies flee you without attacking.
> 
> So you mean, if you have an outpost that belongs to "FriendlyNPCs", and you're spawning NPCs belonging to Team2 inside it, they flee instead of actively fighting you? I'm not entirely sure I'd consider this a bug: it's intentional that by default NPCs flee instead of fighting when they encounter hostiles inside a hostile outpost (unless they've been explicitly ordered to fight enemies). I think this could be addressed by changing the location to belong to Team2 instead, which should make the NPCs consider it "their outpost" and defend against intruders.

The issue is, we're spawning a hostile mission on a friendly outpost so you can fight on said outpost to "liberate" it for your faction. But since this is a friendly outpost and the enemies spawned via the mission are hostile by default, the supposedly hostile AI always switches to flee behavior and we're forced to constantly run CombatActions via looping events.

**Reply by Regalis11** [reply]
I think I'm still confused about what you're trying to do here, or what the expected behavior is. Do you want to have the enemies spawned by the mission to be actively hostile towards the player crew and the outpost NPCs, or just the player crew?

**Reply by ubertpendragon** [reply]
The outpost NPCs are friendly to everyone (FriendlyNPC)
The mission NPCs spawned as Team2 are not hostile towards FriendlyNPCs but should be hostile to players, which they are, **but their AI is broken as they keep fleeing instead of fighting**


We went through several iterations, trying to work around this problem:

Initially we had Team2 + an event forcing a CombatAction loop
Then we removed outpost NPCs entirely
Then we made the mission give you -200 reputation at the start, and spawned the soldiers as FriendlyNPC, so the whole outpost would spot you as hostile (well, except this part https://github.com/FakeFishGames/Barotrauma/discussions/13960)

In an ideal world, we'd just spawn the enemy NPCs in Team2 via mission in the outpost, among the existing residents.

**Comment by itchyOwl** [comment]
Created an issue ticket of this: #13892.

**Comment by Regalis11** [comment]
Addressed in https://github.com/FakeFishGames/Barotrauma-development/commit/1a729305c160a0717376058f92d8118f88f2d839

> Made hostile abandoned outpost NPCs able to react offensively to the player even if the mission takes places in a friendly outpost. Normally hostile NPCs inside "someone else's outpost" retreat from enemies instead of actively fighting them. In the case of NPCs moved to a hostile team by an event, we bypass that by checking if the NPC was originally in the outpost's team - let's do the same for abandoned outpost NPCs.
> 
> Also fixed NPCs in the team None not considering other teams hostile, even though the other teams do consider them hostile.

**Comment by Regalis11** [comment]
This has been fixed in our private repo now, fix coming up in the next unstable build!
